Add Authorization to alertmanager http configs

QuentinBisson opened the pull request:
Closes https://github.com/prometheus-operator/prometheus-operator/issues/3970

This PR fixes the lack of overridability for the authorization token so we could use this to define the Opsgenie GenieKey authorization header

The tests are failing and I'm not sure they are related to my change or if they are flaky


<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:feature
    - release-note:change
    - release-note:none

Unless you choose release-note:none, please add a release note.
-->


**Release Note Template (will be copied)**

```release-note:enhancement
Add Authorization support to alertmanager HTTP config
```

paulfantom commented on pkg/alertmanager/amcfg.go:
```diff
@@ -808,6 +808,19 @@ func (cg *configGenerator) convertHTTPConfig(ctx context.Context, in monitoringv
 		if username != "" && password != "" {
 			out.BasicAuth = &basicAuth{Username: username, Password: password}
 		}
+	} else if in.Authorization != nil {
```
Could we warn when `Authorization` and `BasicAuth` are set?

paulfantom replied:
This is available only in alertmanager 0.22+, so I believe we need a version check.

QuentinBisson replied:
Thanks @paulfantom 
Where can I find an example of a version check ?

Also, does the following log suffice:

```go
if in.BasicAuth != nil && in.Authorization != nil {
		level.Warn(cg.logger).Log("msg", "Basic auth and Authorization are mutually exclusive. Basic auth will take precedence.")
	}
111

paulfantom replied:
Yes, that message looks good :+1: 

Here is an example for version check we do in statefulset.go: https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/alertmanager/statefulset.go#L258-L262

QuentinBisson replied:
Thank you :D

QuentinBisson replied:
I added the check in the operator code as this is the only place we have access to the Alertmanager version

simonpasquier replied:
Side-note: it might make sense to pass the Alertmanager version to `configGenerator`.

QuentinBisson replied:
I can it there as well and not render the authorization part as well

paulfantom commented on Documentation/api.md:
```diff
@@ -1087,7 +1099,8 @@ HTTPConfig defines a client HTTP configuration. See https://prometheus.io/docs/a
 | ----- | ----------- | ------ | -------- |
-| basicAuth | BasicAuth for the client. | *monitoringv1.BasicAuth | false |
+| authorization | Authorization header configuration for the client. This is mutually exclusive with BasicAuth | *monitoringv1.Authorization | false |
+| basicAuth | BasicAuth for the client. This is mutually exclusive with Authorization | *monitoringv1.BasicAuth | false |
```
Would it make sense to deprecate `basicAuth` field?

mig4 commented on Documentation/api.md:
```diff
@@ -235,6 +236,17 @@ ArbitraryFSAccessThroughSMsConfig enables users to configure, whether a service
 
+## Authorization
+
+Authorization allow an endpoint to authenticate by configuring the authentication token More info: https://prometheus.io/docs/operating/configuration/#endpoints
```
```suggestion
Authorization allows configuring authentication with endpoint via a token. More info: https://prometheus.io/docs/operating/configuration/#endpoints
```

simonpasquier commented on pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:
```diff
@@ -490,7 +490,12 @@ func (r *OpsGenieConfigResponder) Validate() error {
+	// +optional
+	Authorization *monitoringv1.Authorization `json:"authorization,omitempty"`
 	// BasicAuth for the client.
+	// This is mutually exclusive with Authorization
```
```suggestion
	// This is mutually exclusive with Authorization. If both are defined, BasicAuth takes precedence.
```

simonpasquier commented on pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:
```diff
@@ -490,7 +490,12 @@ func (r *OpsGenieConfigResponder) Validate() error {
 // See https://prometheus.io/docs/alerting/latest/configuration/#http_config
 type HTTPConfig struct {
+	// Authorization header configuration for the client.
+	// This is mutually exclusive with BasicAuth and is only available starting from alertmanager v0.22+
```
```suggestion
	// This is mutually exclusive with BasicAuth and is only available starting from Alertmanager v0.22+.
```

simonpasquier commented on pkg/assets/types.go:
```diff
@@ -21,6 +21,12 @@ type BasicAuthCredentials struct {
 
+// AuthorizationCredentials represents a credential to be used with
+// authorization header http authentication, see https://tools.ietf.org/html/rfc7617.
+type AuthorizationCredentials struct {
```
why not `type AuthorizationCredentials string`?

QuentinBisson replied:
Makes sense :)

simonpasquier commented on pkg/assets/store_test.go:
```diff
@@ -127,7 +127,7 @@ func TestAddBearerToken(t *testing.T) {
 			}
 
-			key := fmt.Sprintf("basicauth/%d", i)
+			key := fmt.Sprintf("bearertoken/%d", i)
```
:+1:

simonpasquier commented on pkg/alertmanager/operator.go:
```diff
@@ -1361,18 +1362,32 @@ func checkAlertmanagerRoutes(r *monitoringv1alpha1.Route, receivers map[string]s
+	}
+
+	if version.GTE(semver.MustParse("0.22.0")) && httpConfig.Authorization != nil {
+		return errors.Wrap(err, "Authorization is only supported for alertmanager versions greater than v0.22+")
```
I would rather see that the operator is liberal here (e.g. doesn't fail Alertmanager resources that define the `Authorization` field while version < v0.22) and strict when it comes to generate the configuration (e.g. only use the `Authorization` when version >= v0.22). This would match what the existing behavior for the Prometheus CRD.

QuentinBisson replied:
I can do the change but there is no logger down the stack of check methods. Should I just return nil here or put a logger in all checks ? This would just reject the alertmanagerconfig as the code is currently doing for an empty username in basic auth

QuentinBisson replied:
@simonpasquier what do you think I should do regarding this ?

simonpasquier reviewed (commented):
Very nice so far :)

QuentinBisson commented on pkg/alertmanager/operator.go:
```diff
@@ -1361,18 +1362,32 @@ func checkAlertmanagerRoutes(r *monitoringv1alpha1.Route, receivers map[string]s
+		return errors.Wrap(err, "failed to parse alertmanager version")
+	}
+
+	if version.GTE(semver.MustParse("0.22.0")) && httpConfig.Authorization != nil {
```
Yes it does 🤦🏻

QuentinBisson commented on pkg/alertmanager/operator.go:
```diff
@@ -1361,18 +1362,32 @@ func checkAlertmanagerRoutes(r *monitoringv1alpha1.Route, receivers map[string]s
 
+	// Version checks before we configure the http config
 	var err error
+	version, err := semver.ParseTolerant(amVersion)
```
It does :)

QuentinBisson commented on pkg/apis/monitoring/v1/types.go:
```diff
@@ -1039,6 +1039,16 @@ type ProberSpec struct {
 }
 
+// Authorization allow an endpoint to authenticate by configuring the authentication token
+// More info: https://prometheus.io/docs/operating/configuration/#endpoints
```
I took it from basic auth. The correct URL should be https://prometheus.io/docs/alerting/latest/configuration/#http_config

QuentinBisson commented:
@paulfantom could you give it another look ?

QuentinBisson commented:
@simonpasquier would you have time to recheck this PR ? This is currently preventing us from using the Alertmanager

simonpasquier commented on pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:
```diff
@@ -490,7 +490,12 @@ func (r *OpsGenieConfigResponder) Validate() error {
+	// Authorization header configuration for the client.
+	// This is mutually exclusive with BasicAuth and is only available starting from Alertmanager v0.22+.
+	// +optional
+	Authorization *monitoringv1.Authorization `json:"authorization,omitempty"`
```
I'd stick with `SafeAuthorization`

```suggestion
	Authorization *monitoringv1.SafeAuthorization `json:"authorization,omitempty"`
```

simonpasquier commented on pkg/alertmanager/amcfg.go:
```diff
@@ -808,6 +816,19 @@ func (cg *configGenerator) convertHTTPConfig(ctx context.Context, in monitoringv
+		}
+
+		if credentials != "" {
+			authorizationType := out.Authorization.Type
```
```suggestion
			authorizationType := in.Authorization.Type
```

simonpasquier reviewed (commented):
@QuentinBisson sorry for the lag, I was on PTO but back on it now. Regarding https://github.com/prometheus-operator/prometheus-operator/pull/4110#discussion_r654488478, my take is that the operator should gather all Authorization secrets (irrespective of the Alertmanager version) and then log a warning when generating the configuration if the version is < 0.22.0. Similar to what we do for Prometheus:

https://github.com/prometheus-operator/prometheus-operator/blob/83fe36566f4e0894eb5ffcd2638a0f039a17bdeb/pkg/prometheus/promcfg.go#L119-L137

Does it make sense?

QuentinBisson commented:
@simonpasquier everyone has right to their PTOs :D I'm sorry if I was too pushy on this. 
Hopefully, my latest commits should have addressed your reviews :)

simonpasquier commented:
@QuentinBisson would you be able to work on this soon? Otherwise I can take over and address my own comment :)

QuentinBisson commented:
@simonpasquier I came back from my PTO today :D You were right regarding the semver check. Once the build passes, what do you think will be missing here ?

simonpasquier commented:
Cool, hope you had a good PTO time :)
I think we will be good to go then!

simonpasquier commented:
Thanks!

QuentinBisson commented:
Thank you :) Do you known when we can expect the release to plan this on our side ?

simonpasquier commented:
Yes 22th of September is the date planned for v0.51.

QuentinBisson commented:
Thanks :D
